### PR TITLE
[GlobalOptimization] Lift `linalg.generic` ops to `linalg.batch_matmul/linalg.batch_vecmat/linalg.batch_matvec`

### PR DESCRIPTION
**Pull request by Max191**
This PR implements what is discussed in https://github.com/openxla/iree/issues/15214, building off of https://github.com/openxla/iree/pull/15315

### CODE REVIEW

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (Max191) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: @bjacob For BatchVecmat, there are no BatchVecmatTransposeA/B ops, but for BatchMatmul would it be better to add a check here for whether the batch dimension is transposed (and then create the different flavors of BatchMatmulTransposeA/B accordingly)? Or is it okay to leave it as a Transpose + BatchMatmul?

Comment (bjacob) [reply compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Just leave all transpositions in separate Transpose ops. Don't try to generate these special transposing matmul named ops. I don't think they should even exist :-)

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: we usually use `isa<arith::AddFOp, arith::AddIOp>` for the case. 

The l.58 check is redundant if because they are part of add op semantics.

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: let's have a helper function, and use it here. Also, we can use `CastOpInterface` for matching. E.g.,

```cpp
bool isCastOfBlockArgument(Operation *op) {
  return isa<CastOpInterface>(op) && op->getNumOperands() == 1 &&
         isa<BlockArgument>(op->getOperand(0));
}
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Please use the upstream method `invertPermutationVector`:

https://github.com/llvm/llvm-project/blob/77e88db6b7cd982769a852b0fedfc57b3374d1d9/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L228-L229

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: please add a doc

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: You can use the upstream method `applyPermutation`, which is from https://github.com/llvm/llvm-project/blob/77e88db6b7cd982769a852b0fedfc57b3374d1d9/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L217-L226

```suggestion
  SmallVector<int64_t> newInputShape = applyPermutation(inputShape, perm);
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: This does not work for dynamic shapes. Can you support the case?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Let's use `CastOpInterface`. It saves code and makes it cleaner. You can use something similar to below to create the op:

```cpp
        rewriter.create(op->getLoc(), op->getName().getIdentifier(),
                        bcastOp.getSource(), castResTy, op->getAttrs());
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nice check! Maybe we can use `static_assert` since they are compilation time info?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: please use upstream method.

It does not work for dynamic shapes. We should either bail out or support the case

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: you can use `rewriter.getZeroAttr` to get the attribute. I.e., you can simplify it to

```cpp
Value zero = rewriter.create<arith::ConstantOp>(loc, rewriter.getZeroAttr(outType.getElementType()));
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: s/Inputs/Operands

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Please use assignment syntax. The brace initialization with constructor syntax is dangerous. It could potentially lead to a bug, e.g.,

```
// Could invoke an intializer list constructor, or a two-argument constructor.
Frobber frobber{size, &bazzer_to_duplicate};

// Makes a vector of two doubles.
std::vector<double> fifty_pies{50, 3.14};
```

For more context, see https://abseil.io/tips/88

**Review by hanhanW — CHANGES_REQUESTED** [review]
style nit: drop `llvm::` prefix for `llvm::cast`. It's more common in MLIR/IREE codebase.

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: The check is used in several places. Can you refactor it to a function?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: 2023

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: you can merge them into a single `isa`.

```suggestion
  if (!isa<arith::AddFOp, arith::AddIOp>(*addOp)) {
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: remove this debug helper

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: I would like to get an opinion from @hanhanW or @MaheshRavishankar on the match-failure reporting mechanisms. Should we suggest a `rewriter.notifyMatchFailure` call instead ?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Separately from my other question above, I'd like an opinion @hanhanW @MaheshRavishankar on the verbosity of unconditionally logging this to stderr even if it's behind LLVM_DEBUG ?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: For consistency, let us pick one verb and use it consistently: `lift` or `raise` ?

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Let's take a little step back and think of a slightly different interface for these helper functions that will not require duplicating this contraction-dims-matching between the caller (`matchAndRewrite`, figuring which `raise*` helper to call) and the callee (each `raise*` function validating that it was the right helper to call).

Since these helpers return `LogicalResult` already, one way to resolve this would be to replace this `assert` by just a `if(...same condition...) { return failure(); }`.  This way, the caller could just call all of the `raise*` helper functions in sequence until one succeeds. This way, these conditions would become implementation details of these helpers, letting the caller (`matchAndRewrite`) be easier to read.

**Review by bjacob — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Not `conv_2d_input_nhwc_filter_nhwc`

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: Please file an Issue and reference it in a `// TODO(#12345)` comment here about a follow-up addition similar cases for `contractionDims->batch.size() == 0`, creating non-`batch_` ops similar to the `batch_` ops created here.  It's not on our immediate critical path so OK to defer to a TODO, but it'll greatly expand the impact of what you're doing here since in practice there are more matmuls than batch_matmuls outside of the case we're immediately motivated by at the moment.

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: yeah, that would be better. They will be logged to pattern matching debug message, if I remember correctly. Can you change the function to `static LogicalResult hasMatmulBody(RewriterBase &rewriter, linalg::GenericOp genericOp)` and use `rewriter.notifyMatchFailure` here.

Generally, we use `rewriter.notifyMatchFailure` for matching, and `LLVM_DEBUG` for implementation logs.

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: ditto, we can merge it into a single `isa`

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: I think it is fine if the log helps. We have logged similar things in upstream methods (e.g., lowerPack). Maybe can rewrite it using `llvm::interleaveComma`, so you don't have to implement the helper.

```cpp
LLVM_DEBUG({ llvm::interleaveComma(lhsPerm, llvm::dbgs() << "lhsPerm: "); llvm::dbgs() << "\n"; })
```

File: compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp]: nit: use `rewriter.notifyMatchFailure`. Same for the below checks

**Review by hanhanW — APPROVED** [review]
Overall looks okay to me, please wait for Benoit's review.

**Comment by MaheshRavishankar** [comment]
Wondering if there is anything to reuse from here https://github.com/llvm/llvm-project/pull/70326

**Comment by bjacob** [comment]
> Wondering if there is anything to reuse from here https://github.com/llvm/llvm-project/pull/70326

While that is also about specializing generics, it's orthogonal to the present PR in what it actually does, no?

**Comment by hanhanW** [comment]
> > Wondering if there is anything to reuse from here [llvm/llvm-project#70326](https://github.com/llvm/llvm-project/pull/70326)
> 
> While that is also about specializing generics, it's orthogonal to the present PR in what it actually does, no?

I skimmed through the PR and think they are orthogonal. There is nothing we can reuse in this PR.

**Comment by MaheshRavishankar** [comment]
Sorry for the noise then, saw notifications go by for both and they both seem to be doing similar things... didnt look at the details, its a pity if they are not aligned...

**Comment by dcaballe** [comment]
Hey there! Just flying by and curious to know how is that we end up with a generic op that is actually implementing a batch matmul/matvec/vecmat? Thanks!

**Comment by Max191** [comment]
> Hey there! Just flying by and curious to know how is that we end up with a generic op that is actually implementing a batch matmul/matvec/vecmat? Thanks!

This will come from quantized matmuls that get reassociated from https://github.com/openxla/iree/pull/15278

After the reassociation, the number of groups dimension effectively becomes an inner batch dimension, and the main workload is a generalized transposed batch_matmul where the batch dimension has been transposed

**Review by bjacob — APPROVED** [review]

**Comment by Max191** [comment]
Rebased on main, resolved conflicts

**Pull request merged**
